### PR TITLE
Add custom domain git (change git regex)

### LINKED ISSUE #7853: [FR] Add custom domain git
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

In [getGitCommitUrl](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/common/utils/Utils.js#L274-L304) function, it would check only `github.com`, `gitlab.com`.

If I use custom domain git (eg. `git.godhyum.com`), projects's `Source`, `Git Commit` value is not linking.

스크린샷 2023-02-17 오전 11 42 04

So, I want to add custom git option to mlflow.

### Motivation

> #### What is the use case for this feature?
Using custom domain git.
> #### Why is this use case valuable to support for MLflow users in general?
When users use custom domain git, they can go to their git by hypertext link.
> #### Why is this use case valuable to support for your project(s) or organization?
I'm using custom domain git, so I need this option.
> #### Why is it currently difficult to achieve this use case?
MLflow checks only github.com, gitlab.com now.

### Details

Add custom domain git variable like `MLFLOW_GIT_REPO`.

If it exists, set git url to it's value.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [X] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [X] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [X] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by gusghrlrl101**
## Related Issues/PRs

Resolve #7853 

## What changes are proposed in this pull request?

~~Merge `gitHubRegex` and `gitLabRegex` to `gitRegex`.~~
Add new`gitRegex`. (Allow any domain to contain `git`)

## How is this patch tested?
- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by gusghrlrl101** [comment]
[
스크린샷 2023-03-03 오전 11 20 23
](url)

In my case, hypertext link works well in Source, Git Commit.
(Also, it works [mlflow sample](https://github.com/mlflow/mlflow/tree/master/examples/sklearn_elasticnet_wine) in github too)

File: mlflow/server/js/src/common/utils/Utils.js [hunk 1]
```diff
@@ -236,29 +236,23 @@ class Utils {
...
   }
 
-  static getGitHubRegex() {
-    return /[@/]github.com[:/]([^/.]+)\/([^/#]+)#?(.*)/;
-  }
-
-  static getGitLabRegex() {
-    return /[@/]gitlab.com[:/]([^/.]+)\/([^/#]+)#?(.*)/;
+  static getGitRegex() {
+    return /(.*?[@/][^?]*git.*?)[:/]([^#]+)(?:#(.*))?/;
```
Comment (WeichenXu123) [inline mlflow/server/js/src/common/utils/Utils.js:240]: Could you comment some example strings that the regex can match ?

Comment (gusghrlrl101) [reply mlflow/server/js/src/common/utils/Utils.js:248]: I add comment now 🌈

Comment (WeichenXu123) [reply mlflow/server/js/src/common/utils/Utils.js:248]: I suggest split regex for each case matching string,

e.g.,

case 1: `github.com` / `gitlab.com` ,  we should match the whole string of `github` / `gitlab`,

case 2: custom git repo like `https://git.XXX. ...)`, 

This is for better code readability, more strict matching rule, and easier code extensibility.

Comment (gusghrlrl101) [reply mlflow/server/js/src/common/utils/Utils.js]: Thank you for reply!

I added a new regex and left the old ones 🌈

Comment (gusghrlrl101) [reply mlflow/server/js/src/common/utils/Utils.js]: @WeichenXu123 

Hi.
Can you check this?

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @gusghrlrl101 !

**Review by WeichenXu123 — APPROVED** [review]

**Comment by gusghrlrl101** [comment]
Thank you for checking.

I'd wait for merging 🙏🏼

File: mlflow/server/js/src/common/utils/Utils.js [hunk 2]
```diff
@@ -266,6 +278,13 @@ class Utils {
       if (bitbucketMatch[3]) {
         url = url + '/src/master/' + bitbucketMatch[3];
       }
+    } else if (gitMatch) {
+      // eslint-disable-next-line no-unused-vars
+      const [_fullUrl, baseUrl, repoDir, fileDir] = gitMatch;
```
Comment (harupy) [inline mlflow/server/js/src/common/utils/Utils.js:283]:
```suggestion
      const [, baseUrl, repoDir, fileDir] = gitMatch;
```

Comment (harupy) [reply mlflow/server/js/src/common/utils/Utils.js:283]: Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

File: mlflow/server/js/src/common/utils/Utils.js [hunk 3]
```diff
@@ -1032,8 +1062,7 @@ class Utils {
   }
 
   // eslint-disable-next-line prettier/prettier
-  static updatePageTitle(title) {
-  }
+  static updatePageTitle(title) {}
```
Comment (harupy) [inline mlflow/server/js/src/common/utils/Utils.js:1065]:
```suggestion
  static updatePageTitle(title) {
  }
```

to revert a unnecessary change

File: mlflow/server/js/src/common/utils/Utils.js [hunk 4]
```diff
@@ -298,6 +317,17 @@ class Utils {
         sourceVersion +
         '/' +
         bitbucketMatch[3];
+    } else if (gitMatch) {
+      // eslint-disable-next-line no-unused-vars
+      const [_fullUrl, baseUrl, repoDir, fileDir] = gitMatch;
```
Comment (gusghrlrl101) [inline mlflow/server/js/src/common/utils/Utils.js:322]:
```suggestion
      const [, baseUrl, repoDir, fileDir] = gitMatch;
```

@harupy 
Thank you for review.
Maybe I should change this too?

Comment (gusghrlrl101) [inline mlflow/server/js/src/common/utils/Utils.js:322]:
```suggestion
      const [, baseUrl, repoDir, fileDir] = gitMatch;
```
@harupy
Thank you for review.
Maybe I should change this too?

Comment (harupy) [reply mlflow/server/js/src/common/utils/Utils.js]: Thanks for the fix!

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
we can ignore the error

**Comment by gusghrlrl101** [comment]
@harupy 

Thanks 🙏🏼

When will it be reflected in the new version?
(I just want to know release strategy.)
